### ISSUE TITLE
Extreme memory usage when using large ipsets

### ISSUE THREAD

**Issue by principis** [firewalld/firewalld#738]
**What happened**:
I'm using [this guide](https://fedoramagazine.org/protect-your-system-with-fail2ban-and-firewalld-blacklists/) to create a large country blocklist using ipsets.

Adding the ipsets takes a long time (around 5 minutes) causes high cpu usage and steadily increases the memory usage.
When reloading, the memory usage increased to 35%+ (of 8 GB).

After the reload (or restart) firewalld uses 1.5 GB memory (without the ipset it uses 26.7 MB). This seems way too much. 

I tried to remove the ipset, and run `firewall-cmd --reload` after which the ipset was removed, but the memory usage stayed the same, 1.5 GB. This also seems like a bug. Restarting firewalld (without the blacklist) fixes the problem, after which the memory usage dropped to around 26 MB.

**What you expected to happen**:
Not use this much of memory, and reload faster.

**How to reproduce it (as minimally and precisely as possible)**:

```sh
countries="af al am ar bd ca cl cn ee fr gt hk id in kp kr la pl ro ru sg ua um us uy uz ve vg vi vn vu wf ye zm zw"

# Create a temporary working directory
ipdeny_tmp_dir=$(mktemp -d -t blacklist-XXXXXXXXXX)
pushd $ipdeny_tmp_dir

# Download the latest network addresses by country file
curl -LO http://www.ipdeny.com/ipblocks/data/countries/all-zones.tar.gz
tar xf all-zones.tar.gz

firewall-cmd -q --permanent --new-ipset=blacklist --type=hash:net \
    --option=family=inet --option=hashsize=4096 --option=maxelem=200000

# Add the address ranges by country per ipdeny.com to the blacklist
for country in $countries; do
    firewall-cmd -q --permanent --ipset=blacklist \
        --add-entries-from-file=./$country.zone && \
        echo "Added $country to blacklist ipset."
done

firewall-cmd --permanent --zone=drop --add-source=ipset:blacklist
firewall-cmd -q --reload

popd
rm -rf $ipdeny_tmp_dir
```

**Anything else we need to know?**:

**Environment**:
- Firewalld Version: `0.8.4-1.fc32`
- Firewalld Backend: `nftables`
- OS: `Fedora 32 (Server Edition)`
- Others: 
  - nftables: `0.9.3`

**Comment by principis** [comment]
I would like to help and fix/improve this. Is there anything I failed to mention?

**Comment by erig0** [comment]
The memory usage is probably from libnftables. There have been numerous libnftables improvements in the JSON interface lately related to memory and performance. Can you check with a newer libnftables? e.g. 0.9.8

**Comment by afletch** [comment]
Seeing the memory element of this report on CentOS 8.3, with an ipset of ~65k entries about 1.25GB on load and then firewalld running at about 750MB once fully loaded. 
RHEA-2020:4535 (nftables-0.9.3-16) / RHBZ#1834853 with the JSON API fix is applied, hence the CPU issue is not seen.

No such issues when using iptables as the backend.

Environment:
- nftables version: `nftables-0.9.3-16.el8`
- Firewalld version: `firewalld-0.8.2-2.el8`
- firewalld backend: `nftables`
- other
  - `python3-firewall-0.8.2-2.el8`

**Comment by pgnd** [comment]
running

```
grep PRETTY /etc/os-release
	PRETTY_NAME="Fedora 33 (Thirty Three)"
uname -rm
	5.11.15-200.fc33.x86_64 x86_64

rpm -qa | egrep "nft|ipset|firewalld"
	firewalld-filesystem-0.9.3-5.fc33.noarch
	firewalld-0.9.3-5.fc33.noarch
	ipset-libs-7.6-2.fc33.x86_64
	ipset-7.6-2.fc33.x86_64
	libnftnl-1.1.7-3.fc33.x86_64
	nftables-0.9.3-8.fc33.x86_64
	python3-nftables-0.9.3-8.fc33.x86_64
	iptables-nft-1.8.5-6.fc33.x86_64
```

creating a large set
```
firewall-cmd --permanent \
 --delete-ipset=TEST_IPSET

firewall-cmd --reload

firewall-cmd --permanent \
 --new-ipset=TEST_IPSET \
 --type=hash:net \
 --option=family=inet \
 --option=hashsize=32768 \
 --option=maxelem=196608

firewall-cmd --reload

firewall-cmd --add-rich-rule "rule source ipset=TEST_IPSET drop "

firewall-cmd --permanent \
 --ipset=TEST_IPSET \
 --add-entries-from-file=/usr/local/etc/firewalld/test_ipset.ipset

head -n 5 /usr/local/etc/firewalld/test_ipset.ipset
	203.19.111.0/24
	109.73.164.24/30
	62.140.26.208/29
	103.144.89.0/24
	200.14.78.0/24

wc -l /usr/local/etc/firewalld/test_ipset.ipset
	69600 /usr/local/etc/firewalld/test_ipset.ipset

ls -al /usr/local/etc/firewalld/test_ipset.ipset
	-rw-r--r-- 1 root root 1.1M Apr 22 10:10 /usr/local/etc/firewalld/test_ipset.ipset

wc -l /etc/firewalld/ipsets/TEST_IPSET.xml
	69606 /etc/firewalld/ipsets/TEST_IPSET.xml

ls -alt /etc/firewalld/ipsets/TEST_IPSET.xml
	-rw-r--r-- 1 root root 2.2M Apr 22 10:10  /etc/firewalld/ipsets/TEST_IPSET.xml
```

fails on reload
```
firewall-cmd --reload

	ERROR:dbus.proxies:Introspect error on :1.74:/org/fedoraproject/FirewallD1: dbus.exceptions.DBusException: org.freedesktop.DBus.Error.NoReply: Did not receive a reply. Possible causes include: the remote application did not send a reply, the message bus security policy blocked the reply, the reply timeout expired, or the network connection was broken.
	Error: Remote peer disconnected
```

journal shows OOM
```
kernel: oom-kill:constraint=CONSTRAINT_NONE,nodemask=(null),cpuset=/,mems_allowed=0,global_oom,task_memcg=/system.slice/firewalld.service,task=firewalld,pid=3163,uid=0
kernel: Out of memory: Killed process 3163 (firewalld) total-vm:1760964kB, anon-rss:715460kB, file-rss:0kB, shmem-rss:0kB, UID:0 pgtables:2936kB oom_score_adj:0
```

to date, never had oom issues with iptables/ipset loading this -- and other -- large sets.
so seems it's local to firewalld.

**Comment by pgnd** [comment]
updating the F33 pkg version mix to newer/latest
```
firewalld  0.9.3-5.fc33
ipset      7.11-2.fc33
libnftnl   1.1.9-2.fc33
nftables   1:0.9.8-3.fc33
```

doesn't improve the outcome

still
```
time firewall-cmd --reload
	Error: Remote peer disconnected

	real    0m20.698s
	user    0m0.329s
	sys     0m0.078s
```

journalctl -f
...
Apr 22 11:49:27 test.loc kernel: oom-kill:constraint=CONSTRAINT_NONE,nodemask=(null),cpuset=/,mems_allowed=0,global_oom,task_memcg=/system.slice/firewalld.service,task=firewalld,pid=615,uid=0
Apr 22 11:49:27 test.loc kernel: Out of memory: Killed process 615 (firewalld) total-vm:1750588kB, anon-rss:726412kB, file-rss:0kB, shmem-rss:0kB, UID:0 pgtables:2924kB oom_score_adj:0
Apr 22 11:49:27 test.loc systemd[1]: firewalld.service: A process of this unit has been killed by the OOM killer.
Apr 22 11:49:27 test.loc kernel: oom_reaper: reaped process 615 (firewalld), now anon-rss:0kB, file-rss:0kB, shmem-rss:0kB
Apr 22 11:49:27 test.loc systemd[1]: firewalld.service: Main process exited, code=killed, status=9/KILL
Apr 22 11:49:27 test.loc systemd[1]: firewalld.service: Failed with result 'oom-kill'.
```

**Comment by pgnd** [comment]
> Can you check with a newer libnftables? e.g. 0.9.8
>> erig0 added the needinfo label on Feb 18 

@erig0 ^^

**Comment by SkewedZeppelin** [comment]
I can reproduce this under Fedora 34 with Linux 5.11.17, firewalld 0.9.3, nftables 0.9.8.
I think it might be related to auditing of NETFILTER_CFG nft_register_rule by SELinux policy.
I've seen small lists quickly surpass 30+GB of RAM and take down the system.

Here is a simple reproducer:
```sh
wget https://iplists.firehol.org/files/firehol_level4.netset;
firewall-cmd --new-ipset=firehol_level4 --type=hash:net --option=maxelem=1000000;
firewall-cmd --ipset=firehol_level4 --add-entries-from-file=firehol_level4.netset;
firewall-cmd --zone=drop --add-source=ipset:firehol_level4;
```

**Comment by pgnd** [comment]
fwiw, selinux is completely DISabled on my system

**Comment by pgnd** [comment]
as I currently understand:

firewalld has an option for 'nftables' backend

if backend = nftables, much of the generated firewall uses nftables' native constructs.

but nftsets are neither used, nor directly available.  firewalld's use of ipsets embeds the ipsets in the nft ruleset -- and does not translate to nftsets.

for small ipsets, shouldn't be a problem -- tho i'm unclear at exactly which which point a list stops being 'small', and this issue^^ crops up.

for large lists -- particularly in the *early* blocklists use case -- i've mv'd all my lists into native nftsets, in an  netdev ingress chain.

that chain inits/loads, and tears down, with firewalld systemd start, but is otherwise managed externally to firewalld.
firewalld is not involved in their creation/editing/destruction at all.

i assume/hope that at some point, firewalld gets support for native nftsets, and this issue^^ resolve.

until that time, in this mode, i avoid the libnftables issues^^. no problems with list (re)loads, so far.
large lists are cleanly/efficiently handled in native nft, and the rest enjoys the convenience of the firewalld frontend.

**Comment by erig0** [comment]
firewalld _does_ translate the "ipset" into a native nftables set.

**Comment by pgnd** [comment]
> firewalld does translate the ipset into a native nftables set

Then I don't see the whole picture.

Can you provide an example of using firewalld-cmd to get a (small, for now) ipset-in-a-file into a native nftables set, in a netdev-family table, rule-addressable in an ingress chain?

**Comment by erig0** [comment]
> > firewalld does translate the ipset into a native nftables set
> 
> Then I don't see the whole picture.
> 
> Can you provide an example of using firewalld-cmd to get a (small, for now) ipset-in-a-file into a native nftables named set, in a netdev-family table, rule-addressable in an ingress chain?

firewalld does not use the `netdev` family. So it won't add them there. Current code only uses the `inet` family. Older code additionally uses the `ip` and `ip6` families.

**Comment by erig0** [comment]
Example:
```
# firewall-cmd --permanent --new-ipset foobar --type hash:net
success
# firewall-cmd --permanent --ipset foobar --add-entry 10.0.1.0/24
success
# firewall-cmd --reload
success
# nft list set inet firewalld foobar
table inet firewalld {
        set foobar {
                type ipv4_addr
                flags interval
                elements = { 10.0.1.0/24 }
        }
}
```

**Comment by pgnd** [comment]
> firewalld does not use the netdev family

_eventually_, would be helpful to be able to specify family & chain hook type options @ firewall-cmd , to both cover all available types & to explicitly limit which chains rules are created in

for my use case, til then, external nftset appears to be the cleanest, if not only, option atm.

**Comment by Bill-ColeUS** [comment]
I also have a large set of ipset blocklists and memory usage skyrockets to several GB. Currently affected systems that I tested are:

- Fedora 32, 33, 34
- CentOS 8.0, 8.1, 8.2, 8.3, 8.4

Any luck getting this fixed?

**Comment by jaboutboul** [comment]
Seems like users are seeing this as well on AlmaLinux 8.4. We've gotten at least a couple of reports. Here is one from our forums: https://almalinux.discourse.group/t/firewalld-using-too-much-memory-alma-8-4/276

**Comment by principis** [comment]
Just tried it again with Firewalld 0.9.3, nftables 0.9.8 and a blacklist of around 20500 entries.


Memory usage was checked every second with systemctl status:
```
     Memory: 24.7M
     Memory: 24.7M
     Memory: 24.7M
     Memory: 25.6M
     Memory: 26.1M
     Memory: 31.4M
# reload firewalld
     Memory: 84.4M
     Memory: 151.3M
     Memory: 456.6M
     Memory: 456.6M
     Memory: 456.6M
     Memory: 456.6M
     Memory: 456.6M
     Memory: 456.6M
     Memory: 456.6M
     Memory: 456.6M
     Memory: 456.6M
     Memory: 456.6M
     Memory: 456.6M
     Memory: 456.6M
     Memory: 456.6M
     Memory: 456.6M
     Memory: 456.6M
     Memory: 456.6M
     Memory: 456.6M
     Memory: 456.6M
     Memory: 456.6M
     Memory: 456.6M
     Memory: 456.6M
     Memory: 456.6M
     Memory: 456.6M
     Memory: 456.6M
     Memory: 456.6M
     Memory: 456.6M
     Memory: 457.6M
     Memory: 307.6M
     Memory: 307.6M
     Memory: 307.6M
```

**Comment by Bill-ColeUS** [comment]
@principis So if I guess correctly, your ~2MB blacklist results in 456MB of maximum memory use? That is just terrible. No wonder my CentOS 8 firewalld uses several GB.

Try this, in a test CentOS 7 system, load the same blacklist, you'll be amazed at how much less memory firewalld uses.

So is this a... 

- Memory leak?
- Design issue?
- Bug?
- Feature?  :smile:

**Comment by erig0** [comment]
> @principis So if I guess correctly, your ~2MB blacklist results in 456MB of maximum memory use? That is just terrible. No wonder my CentOS 8 firewalld uses several GB.
> 
> Try this, in a test CentOS 7 system, load the same blacklist, you'll be amazed at how much less memory firewalld uses.

You can get the same results on modern versions by setting `FirewallBackend=iptables`. The cause of the memory usage was indicated above if anyone cares to read the entire thread.

**Comment by principis** [comment]
@erig0 sorry, my mistake. 

According to your earlier post, the issue is how the ipset is processed. That seems strange, because the memory usage is overall very high. So I looked at the code, and the issue seems to be the [rule_ref_count](https://github.com/firewalld/firewalld/blob/master/src/firewall/core/nftables.py#L169) and [rule_to_handle](https://github.com/firewalld/firewalld/blob/master/src/firewall/core/nftables.py#L169) dictionaries. They become massive with a very large ipset and are kept in memory at all times.

**Comment by Bill-ColeUS** [comment]
So this firewall is a firewall that can't block a lot of IP addresses... lol

Shouldn't this be of higher priority? Firewalld is supposed to be a firewall... right?  :roll_eyes:

**Comment by erig0** [comment]
> So this firewall is a firewall that can't block a lot of IP addresses... lol
> 
> Shouldn't this be of higher priority? Firewalld is supposed to be a firewall... right? roll_eyes

Patches are welcome. Complaining is not.

**Comment by pgnd** [comment]
No, firewalld is an enduser UI/mgmt tool.  It's a frontend to kernel's nftables framework, and supports iptables & nftables backends.

This^^ is an issue with libnftables' memory mgmt when passed large data sets.

If you need large data sets, create them as nfsets directly with nft , and load the tables in parallel with firewalld's.

**Comment by erig0** [comment]
See related RHEL bug: https://bugzilla.redhat.com/show_bug.cgi?id=1908127
This is without firewalld being involved.

**Comment by Bill-ColeUS** [comment]
As a temporary workaround, is it enough to just set `FirewallBackend=iptables` in the .conf file?

Thank you.

**Comment by erig0** [comment]
> As a temporary workaround, is it enough to just set `FirewallBackend=iptables` in the .conf file?

yes.

**Comment by cryptoluks** [comment]
Answering https://github.com/firewalld/firewalld/issues/849#issuecomment-911548031 here, as it is probably more related to this issue.

> There is an issue with memory pressure and large ipsets. It's possible this is causing paging and really slow performance. See #738.

Thanks for the hint.

I tried setting `FirewallBackend=iptables` and restarted `firewalld`. However, adding my 18k subnets to an ipset did not work with the `iptables` backend either.

I am using openSUSE MicroOS with BTRFS snapshots and just rolled back the OS to around 4th August. Then, the ipset was built in a matter of seconds again. I did not pinpoint the exact version yet, but with `libnftables 0.9.9` and `firewalld 0.9.3` on my older OS snapshot the issue went away. Before I was on `libnftables 1.0.0` and `firewalld 1.0.0`.

So I will probably stick to the older versions for now.

**Comment by wwalker** [comment]
I am also seeing this problem. 

What takes less than 1 second on CentOS 7, takes longer than 15 minutes (when I gave up, since I have 20 fedora 35 machines to add these rules to) with either nftables or iptables backend.

There are 13,119 lines in my cidr file.
```
 /usr/bin/firewall-cmd --permanent --ipset=bad-people --add-entries-from-file=bad-people.cidr
```

`top` says:
```
    PID USER      PR  NI    VIRT    RES    SHR S  %CPU  %MEM     TIME+ COMMAND
 431036 root      20   0  272372  39084  15168 R 100.0   0.1  18:16.95 firewall-cmd
```

**Comment by Bill-ColeUS** [comment]
Welcome to the club...

We are waiting of RHEL 8.6 to see if the bug is fixed properly or if we are going to stick with iptables for the long term.

**Comment by erig0** [comment]
> I am also seeing this problem.
> 
> What takes less than 1 second on CentOS 7, takes longer than 15 minutes (when I gave up, since I have 20 fedora 35 machines to add these rules to) with either nftables or iptables backend.
> 
> There are 13,119 lines in my cidr file.
> 
> ```
>  /usr/bin/firewall-cmd --permanent --ipset=bad-people --add-entries-from-file=bad-people.cidr
> ```

This was fixed via #881 and #907. It is not the same thing as this issue.

**Comment by wwalker** [comment]
@erig0 is right regarding my problem, 1.0.3 just came out for fedora 35 and has the fixes mentioned above, and now it only takes 2 minutes to create a 13000 entry ipset, instead of infinity :-)

Thank you @erig0

**Comment by afletch** [comment]
I wouldn't say 2mins for 13k compares well with a few seconds for 4x that number of entries using the iptables backend

**Comment by erig0** [comment]
> @erig0 is right regarding my problem, 1.0.3 just came out for fedora 35 and has the fixes mentioned above, and now it only takes 2 minutes to create a 13000 entry ipset, instead of infinity :-)

v1.0.3 contained #881, but not #907 which should take it down to the order of seconds.
v1.0.4 will contain #907. v1.0.4 is scheduled as part of the normal quarterly release (e.g. April 1st-ish).

v1.1.0 already containers #907.

**Comment by Bill-ColeUS** [comment]
Unfortunately, all of the above fixes are rather useless to RHEL (and derivatives, like Alma/Rock) users, still stuck with firewalld 0.9.3.

It may take a few years until the fixes trickle down to us :(

**Comment by erig0** [comment]
> Unfortunately, all of the above fixes are rather useless to RHEL (and derivatives, like Alma/Rock) users, still stuck with firewalld 0.9.3.
> 
> It may take a few years until the fixes trickle down to us :(

Maybe you should file an issue against RHEL.

**Comment by BlackMage2** [comment]
ehm why the distributions stuck at firewalld 0.93?

**Comment by Dinfangion** [comment]
re-tested on AlmaLinux release 8.6 (Sky Tiger) 
not fixed

it has:
Name         : firewalld
Version      : 0.9.3
Release      : 13.el8
Source       : firewalld-0.9.3-13.el8.src.rpm

**Comment by ZIT2050** [comment]
Hello guys, thanks for your discussion and help. My observations at 2022-10-26, the same iptset hash:net 200k, loading IPs from file.

#### Centos 7.9.2009 / kernel=3.10.0-1160.76.1.el7.x86_64
rpm -qa | egrep "nft|firewall|iptab"
firewalld-filesystem-0.6.3-13.el7_9.noarch
iptables-1.4.21-35.el7.x86_64
python-firewall-0.6.3-13.el7_9.noarch
firewalld-0.6.3-13.el7_9.noarch
// Only iptables available in this OS version
// Works fine and fast <1s

#### Rocky Linux  8.6 /  kernel=4.18.0-372.26.1.el8_6.x86_64
rpm -qa | egrep "nft|firewall|iptab"
libnftnl-1.1.5-5.el8.x86_64
nftables-0.9.3-25.el8.x86_64
iptables-1.8.4-22.el8.x86_64
iptables-libs-1.8.4-22.el8.x86_64
iptables-ebtables-1.8.4-22.el8.x86_64
python3-nftables-0.9.3-25.el8.x86_64
python3-firewall-0.9.3-13.el8.noarch
firewalld-filesystem-0.9.3-13.el8.noarch
firewalld-0.9.3-13.el8.noarch
// With nftables generates OOM
// Change backend to iptables HELPS, working good but sometimes problem with reload

#### Rocky Linux 9.0 / kernel=5.14.0-70.26.1.el9_0.x86_64
 rpm -qa | egrep "nft|firewall|iptab"
libnftnl-1.1.9-4.el9.x86_64
nftables-0.9.8-12.el9.x86_64
iptables-libs-1.8.7-28.el9.x86_64
iptables-nft-1.8.7-28.el9.x86_64
python3-nftables-0.9.8-12.el9.x86_64
python3-firewall-1.0.0-4.el9.noarch
firewalld-filesystem-1.0.0-4.el9.noarch
firewalld-1.0.0-4.el9.noarch
// With nftables generates OOM
// Change backend to iptables DON'T HELP, loading into ipsets with hours/days


#### Others:
In Rocky Linux 8.6 and 9.0 found new problem
Error: INVALID_ENTRY: Entry '2.16.0.0/13' overlaps entry '2.16.172.0/24'

old behavior = skip and load ==> ipset is not empty
current behavior = error and exit ==> ipset is EMPTY!

**Comment by Talisker69** [comment]
Hi alls
I reproduced the problem on fedora 37 with nftables on background.
Still no workaround ?

**Comment by uhlhosting** [comment]
Guys, what is going on? Such issues should be addressed asap, killing peoples RAM is not fun!

**Comment by uhlhosting** [comment]
@ZIT2050 already the settings is on iptables in the .conf file. I do not use subnets. Pure IPs.

**Comment by erig0** [comment]
I tried to fix this is #1106 by chunking the entries, but the memory usage still grows. Perhaps there is something going on in libnftables that prevents it from being released or reused.

**Comment by erig0** [comment]
Latest update seems to address the problem. With #1106 the memory is mannered to about 128MB in the given test case. Without chunking it consumes 500+ MB.

**Comment by erig0** [comment]
> Should we treat this as a workaround only? Maybe we should contact netfilter core team.

No. It's a fix. The growing memory problem I mentioned a couple comments up is probably just due to python's garbage collection.

> How to test this, on which systems which packages versions etc.

I'm due to cut stable releases this week. So just wait for those. :)

**Comment by erig0** [comment]
This has been included in releases `v1.2.5` an `v1.3.1`. Thanks for your patience.
